### ISSUE TITLE
postfix-sasl ban failure

### ISSUE THREAD

**Issue by dciwill** [fail2ban/fail2ban#2530]
### Environment:

- Fail2Ban version (including any possible distribution suffixes): Fail2Ban v0.9.7
- OS, including release name/version: CentOS Linux release 7.7.1908 (Core)
- [x] Fail2Ban installed via OS/distribution mechanisms
- [x] You have not applied any additional foreign patches to the codebase
- [x] Some customizations were done to the configuration (provide details below is so)

### The issue:
Failed to execute ban jail 'postfix-sasl' action 'route'

#### Steps to reproduce
Enable fail2ban jail for postfix-sasl and fail the authorization

#### Expected behavior
Ban the IP in route

#### Observed behavior
Error message

#### Any additional information
There is a similar issue at https://github.com/fail2ban/fail2ban/issues/2092 to which the solution is to replace "imap3" with "imap". Incidentally, this was true in this case and I did perform the edit then reloaded the configuration, but that did not fix the issue.

### Configuration, dump, and other helpful excerpts
[https://pastebin.com/bQBYsaL4](https://pastebin.com/bQBYsaL4)

#### Any customizations done to /etc/fail2ban/ configuration
Only enabled route instead of iptables, enabled the jail in jail.local and set maxretry = 2

#### Relevant parts of /var/log/fail2ban.log file:
_preferably obtained while running fail2ban with `loglevel = 4`_
```
2019-09-19 01:59:20,480 fail2ban.actions        [15248]: ERROR   Failed to execute ban jail 'postfix-sasl' action 'route' info 'CallingMap({'ipjailmatches': <function <lambda> at 0x7f60a97b3f50>, 'matches': '2019-09-19T01:57:01.067776mail.REDACTED.com postfix/smtpd[15203]: warning: unknown[46.38.144.32]: SASL LOGIN authentication failed: UGFzc3dvcmQ6\n2019-09-19T01:59:20.034983mail.REDACTED.com postfix/smtpd[15136]: warning: unknown[46.38.144.32]: SASL LOGIN authentication failed: UGFzc3dvcmQ6', 'ip': '46.38.144.32', 'ipmatches': <function <lambda> at 0x7f60a97b3cf8>, 'ipfailures': <function <lambda> at 0x7f60a97b3e60>, 'time': 1568883560.474884, 'failures': 2, 'ipjailfailures': <function <lambda> at 0x7f60a97b3ed8>})': Error banning 46.38.144.32
```
#### Relevant lines from monitored log files in question:

I don't know where to find this.

**Comment by sebres** [comment]
Is this (`Failed to execute ban jail 'postfix-sasl' action 'route' ... Error banning`) the singe error only?
Can you find something helpful in log before (or after) this line?
Note that 0.9 is too old, I don't know in which constellation it'd log something additionally (0.10 would log also script/command-line caused the error).
Have no possibility to open your excerpts from pastebin (blocked on my side atm), you can also add attachment or file, or put it into comment (between \`\`\`new-line ... new-line\`\`\`).

**Comment by dciwill** [comment]
I encountered more errors as the jails for these retried the bans.

Here is the jail.local file I tried to pastebin.
[jail.local.txt](https://github.com/fail2ban/fail2ban/files/3631807/jail.local.txt)

Here is the entire log file:
[fail2ban.log](https://github.com/fail2ban/fail2ban/files/3631849/fail2ban.log)

The additional errors are like:
2019-09-19 00:54:23,674 fail2ban.action         [3569]: ERROR   ip route add unreachable 46.38.144.146 -- stdout: ''
2019-09-19 00:54:23,674 fail2ban.action         [3569]: ERROR   ip route add unreachable 46.38.144.146 -- stderr: 'RTNETLINK answers: File exists\n'
2019-09-19 00:54:23,674 fail2ban.action         [3569]: ERROR   ip route add unreachable 46.38.144.146 -- returned 2
2019-09-19 00:54:23,675 fail2ban.actions        [3569]: ERROR   Failed to execute ban jail 'postfix-sasl' action 'route' info 'CallingMap({'ipjailmatches': <function <lambda> at 0x7f7211ae1e60>, 'matches': '2019-09-19T00:51:24.452066mail.REDACTED.com postfix/smtpd[3496]: warning: unknown[46.38.144.146]: SASL LOGIN authentication failed: UGFzc3dvcmQ6\n2019-09-19T00:52:45.335687mail.REDACTED.com postfix/smtpd[3496]: warning: unknown[46.38.144.146]: SASL LOGIN authentication failed: UGFzc3dvcmQ6\n2019-09-19T00:54:04.018142mail.REDACTED.com postfix/smtpd[3496]: warning: unknown[46.38.144.146]: SASL LOGIN authentication failed: UGFzc3dvcmQ6', 'ip': '46.38.144.146', 'ipmatches': <function <lambda> at 0x7f7211ae1ed8>, 'ipfailures': <function <lambda> at 0x7f720bb2e050>, 'time': 1568879663.637859, 'failures': 3, 'ipjailfailures': <function <lambda> at 0x7f720bb2e0c8>})': Error banning 46.38.144.146

**Comment by sebres** [comment]
> RTNETLINK answers: File exists

This issue seems to be duplicate of https://github.com/fail2ban/fail2ban/issues/2373#issuecomment-473714585 (and originated by impossibility to use some anchor or unique chain name using `ip route` (at least how it is implemented at the moment).

Thus closed as duplicate, please further in #2373.

**Comment by sebres** [comment]
Short excerpt illustrating the issue:
```
NOTICE  [dovecot] Ban 185.234.218.156
NOTICE  [postfix-sasl] Ban 185.234.218.156
ERROR   ip route add unreachable 185.234.218.156 -- stderr: 'RTNETLINK answers: File exists\n'
```
Firstly IP is banned in `dovecot`, hereafter wanted to ban same IP in `postfix-sasl`, what is impossible due to reasons described in mentioned issues.

**Comment by dciwill** [comment]
If IP is already added to route, then why is postfix-sasl receiving a connection from it at all? Doesn't this indicate that "ip route add unreachable" isn't working?

**Comment by sebres** [comment]
It is not necessarily the case (and I don't see it in your log) - there could be several "expected" reasons for that:
- (more probable here) dovecot uses postfix as an authentication interface, so the failure will be simply found in both logs (because both are affected)
- the unauthorized requests (connections) took place on both services for different ports;
- both services were affected at the same time (or due to certain latency between finding and banning there are still messages in log/journal), etc pp.

Anyway it works basically if you don't see too much `... already banned` (and `Found ...`) messages, which occurred several seconds (or even minutes) after IP going banned (but also ignore such notices by start of fail2ban, because it banning during analyze of old messages from log, v.0.10 does not produce "already banned" for such messages).

---

The workaround could be extending such actions (e. g. also shorewall, see #2031) with something like `--if-not-exists` or providing `ignorecommand` which could check the IP is already banned (assuming it is possible) and ignore 2nd bans (and annoying errors in fail2ban.log).

The main issues beginning when your affected jails have for example different `bantime` (so an unban would take place earlier where in another jail it is still "banned", so new ban does not take place (and you'd catch then `already banned`). Or if such ban-actions are multi-port actions (so should normally protect and ban on different ports only).

Better would be to use some different table_id (number) instead of `unreachable` type, that should be pre-configured to route all to `unreachable` (something in `/etc/iproute2/rt_tables`) or to use VRF or something else as anchor. 
Or (if available) to force route-action to count the IP across jails.

"Good" actions are supporting table/chain names or anchors, and would just use different tables for that, so the tickets don't "intersect" there.
"Bad" is that although we have theoretical possibility to fix it within inside of fail2ban ban-manager, but it's a bit complex because it was never intended to work across jails (all jails have its own ban-manager). In version 0.11 we have a common observer, which can be used for that, but just too few time (and many other issues and todos with higher priorities), especially because you can use alternate ban action.
